hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1064 @@
+          } catch (HoodieIOException e) {
+            HoodieInstant cleanInstant = new HoodieInstant(false, instant.getAction(), instant.getTimestamp());
+            ValidationUtils.checkArgument(cleanInstant.isCompleted());
+            HoodieCleanMetadata cleanMetadata = CleanerUtils.getCleanerMetadata(datasetMetaClient, cleanInstant);
+            update(cleanMetadata, instant.getTimestamp());
+          }
+          break;
+        }
+        case HoodieTimeline.DELTA_COMMIT_ACTION:
+        case HoodieTimeline.COMMIT_ACTION:
+        case HoodieTimeline.COMPACTION_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          HoodieCommitMetadata commitMetadata = HoodieCommitMetadata.fromBytes(
+              timeline.getInstantDetails(instant).get(), HoodieCommitMetadata.class);
+          update(commitMetadata, instant.getTimestamp());
+          break;
+        }
+        case HoodieTimeline.ROLLBACK_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          HoodieRollbackMetadata rollbackMetadata = TimelineMetadataUtils.deserializeHoodieRollbackMetadata(
+              timeline.getInstantDetails(instant).get());
+          update(rollbackMetadata, instant.getTimestamp());
+          break;
+        }
+        case HoodieTimeline.RESTORE_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          HoodieRestoreMetadata restoreMetadata = TimelineMetadataUtils.deserializeHoodieRestoreMetadata(
+              timeline.getInstantDetails(instant).get());
+          update(restoreMetadata, instant.getTimestamp());
+          break;
+        }
+        case HoodieTimeline.SAVEPOINT_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          // Nothing to be done here
+          break;
+        }
+        default: {
+          throw new HoodieException("Unknown type of action " + instant.getAction());
+        }
+      }
+    }
+  }
+
+  /**
+   * Return an ordered list of instants which have not been synced to the Metadata Table.
+
+   * @param datasetMetaClient {@code HoodieTableMetaClient} for the dataset
+   */
+  public List<HoodieInstant> findInstantsToSync(HoodieTableMetaClient datasetMetaClient) {
+    HoodieTimeline metaTimeline = metaClient.reloadActiveTimeline();
+
+    // All instants since the last time metadata table was compacted are candidates for sync
+    Option<String> compactionTimestamp = getLatestCompactionTimestamp();
+
+    // If there has not been any compaction then the first delta commit instant should be the one at which
+    // the metadata table was created. We should not sync any instants before that creation time.
+    Option<HoodieInstant> oldestMetaInstant = Option.empty();
+    if (!compactionTimestamp.isPresent()) {
+      oldestMetaInstant = Option.fromJavaOptional(metaTimeline.getInstants()
+          .filter(i -> i.isCompleted() && i.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION))
```

Comment:
metaTimeline.getDeltaCommitTimeline().filterCompletedInstants()? instead of writing this by hand

Reply:
Done.